cheevos-new/cheevos.c

```diff
@@ -1545,6 +1565,186 @@ static int rcheevos_iterate(rcheevos_coro_t* coro)
       MD5_Final(coro->hash, &coro->md5);
       CORO_GOTO(RCHEEVOS_GET_GAMEID);
 
+      /**************************************************************************
+       * Info   Tries to identify a Playstation game
+         * Input  CHEEVOS_VAR_INFO the content info
+         * Output CHEEVOS_VAR_GAMEID the Retro Achievements game ID, or 0 if not found
+         *************************************************************************/
+   CORO_SUB(RCHEEVOS_PSX_MD5)
+   {
+      ext = path_get_extension(coro->path);
+
+      MD5_Init(&coro->md5);
+
+      /* if we're looking at an m3u file, get the first disc from the playlist */
+      if (string_is_equal_noncase(ext, "m3u"))
+      {
+         coro->cdrom_stream = intfstream_open_file(coro->path, RETRO_VFS_FILE_ACCESS_READ, RETRO_VFS_FILE_ACCESS_HINT_NONE);
+         if (coro->cdrom_stream)
+         {
+            char disc_path[PATH_MAX_LENGTH];
+            char* tmp;
+
+            intfstream_read(coro->cdrom_stream, buffer, sizeof(buffer));
+            intfstream_close(coro->cdrom_stream);
+            CHEEVOS_FREE(coro->cdrom_stream);
+            coro->cdrom_stream = NULL;
+
+            tmp = buffer;
+            while (*tmp && *tmp != '\n')
+               ++tmp;
+            if (tmp > buffer && tmp[-1] == '\r')
+               --tmp;
+            *tmp = '\0';
+
+            fill_pathname_basedir(disc_path, coro->path, sizeof(disc_path));
+            strlcat(disc_path, buffer, sizeof(disc_path));
+
+            free((void*)coro->path);
+            coro->path = strdup(disc_path);
+            ext = path_get_extension(coro->path);
+         }
+      }
+
```

Comment:
This is where the primary feature begins.
* open cd track 1
* find SYSTEM.CNF
* read BOOT= line from SYSTEM.CNF
* open file pointed to by BOOT=
* hash that